cpp/src/Slice/Parser.cpp

```diff
@@ -4222,29 +4191,25 @@ Slice::Operation::createParameter(const string& name, const TypePtr& type, bool
         return nullptr;
     }
 
-    // Check that in parameters don't follow out parameters.
-    if (!_outParameters.empty() && !isOutParam)
+    if (isOutParam)
     {
+        _usesOutParameters = true;
+    }
+    else if (_usesOutParameters)
+    {
+        // In parameters must be declared before out parameters.
         _unit->error("`" + name + "': in parameters cannot follow out parameters");
     }
```

Comment:
Can we get rid of `_usesOutParameters` and check if `_returnValues` is empty?

```suggestion
    if (!isOutParam && !_returnValues.empty())
    {
        // In parameters must be declared before out parameters.
        _unit->error("`" + name + "': in parameters cannot follow out parameters");
    }
```

Reply:
Not necessarily, _returnValues contains both out parameters and the return types. So if there's a return type, this would trigger a false error (since _returnValues.empty() would be false, even though there's no out parameters in it)